### ISSUE TITLE
ability to detect when a key is down

### ISSUE THREAD

**Issue by ajloveslily14** [gnembon/scarpet#25]
it may be cool to have an event or something for when a player presses a key(maybe just a key bind or even maybe all keys) that way you could create things like a fast forward key. while you hold down the key, the games runs as fast as it can

**Comment by gnembon** [comment]
Good suggestion. I would need to check if that would only be available in singleplayer or on the server as well, but that would be very useful especially with the hooks system I have planned for v2.

**Comment by ajloveslily14** [comment]
Yeah i could see how doing that in multiplayer could be a bit difficult, because as it is now there wouldn't be any way to tell _who_ pressed a key, just that it was pressed

**Comment by fxmorin** [comment]
Didn’t know that servers received all key presses from client. (Unless they actually don’t) If they do, a bit of Mixin to log the player name should be possible.

**Comment by ajloveslily14** [comment]
I have no idea if they do, I'd assume they don't. But it is minecraft we're talking about

**Comment by PegasusEpsilon** [comment]
@fxmorin @ajloveslily14 (gnembon ought to already know this, having his nose in the code...)

99.99999% sure the client does not send *any* key events, at all, to the server. Only gameplay events like move packets, interaction events, etc. Thus, this would require a client mod, and it's why bukkit plugins make such heavy use of crouching.

About all the data you can count on receiving, from a stock client, is crouch, jump, move, look, use tool, tile entity interact, mob entity interact, drop item, pick up item, attack (take a swing, whether you hit anything or not), take damage, deliver damage (successfully struck mob entity), break block start, break block finish, and place block. Oh, and chat messages, including commands that the client doesn't know how to handle.

The base game doesn't need anything more, so the vanilla client doesn't send anything more. There are some fairly outdated protocol specs published out there, if you're interested, that will tell you more or less exactly what minecraft's client and server say to each other, about 10-15 versions ago.

**Comment by ajloveslily14** [comment]
Does make sense that the client wouldn't send keys. *however* maybe it could be a only-works-with-carpet-client thing. Then maybe a function to test if a player is running the carpet client, if they are then use the custom keybind stuff, else just use crouch or whatever.

**Comment by fxmorin-seneca** [comment]
This seems to go against what Gnembon is trying to do. He wants this to be more of a spigot type deal then forge. Where it's server-side without the need for client-side. Although it would be nice to check how it works.

**Comment by PegasusEpsilon** [comment]
I don't think enabling extra functionality between a scarpet client and a scarpet server technically goes against the philosophy, but I **do** think it would lead to scripts that would go against the philosophy, in that, once you have this ability, you'll use it to write scripts, and those scripts will _require_ you to have a scarpet client to function properly (or even at all), which breaks the philosophy right there.

So basically, vanilla doesn't support it, a mod (like scarpet) could, but given scarpet's philosophy (be like bukkit, not like forge) it shouldn't. And I'm good with that, personally. Vanilla only needs like five more things, max, to be fully server-side moddable with a pure vanilla client. Nag mojang to add them, imo.
